source/client/factories_impl.cc

```diff
@@ -172,15 +174,19 @@ TerminationPredicateFactoryImpl::TerminationPredicateFactoryImpl(const Options&
 TerminationPredicatePtr
 TerminationPredicateFactoryImpl::create(Envoy::TimeSource& time_source, Envoy::Stats::Scope& scope,
                                         const Envoy::MonotonicTime scheduled_starting_time) const {
-  TerminationPredicatePtr duration_predicate = std::make_unique<DurationTerminationPredicateImpl>(
-      time_source, options_.duration(), scheduled_starting_time);
-  TerminationPredicate* current_predicate = duration_predicate.get();
+  TerminationPredicatePtr root_predicate = std::make_unique<NullTerminationPredicateImpl>();
+  if (options_.duration() > 0s) {
+    root_predicate = std::make_unique<DurationTerminationPredicateImpl>(
```

Comment:
Readability wise - it may be better to have an explicit if<->else rather then set a value and modify it after. The latter may confuse the reader if they don't notice the modification under the if statement.

Consider (pseudo code):
root_predicate = NULL
if (finite_run) {
  root_predicate = TerminateAt...
} else {
  root_predicate = NeverTerminate
}